packages/core/types/index.js

```diff
@@ -462,32 +497,62 @@ export type DependencyOptions = {|
  * @section transformer
  */
 export interface Dependency {
+  /** The id of the dependency. */
   +id: string;
-  /** E.g. "lodash" in <code>import {add} from "lodash";</code>  */
-  +moduleSpecifier: ModuleSpecifier;
-  +isAsync: boolean;
-  /** Whether this should become a entry in a bundle. */
-  +isEntry: ?boolean;
-  /** Whether a failed resolution should not cause a build error. */
+  /** The specifier used to resolve the dependency. */
+  +specifier: DependencySpecifier;
+  /**
+   * How the specifier should be interpreted.
+   *   - esm: An ES module specifier. It is parsed as a URL, but bare specifiers are treated as node_modules.
+   *   - commonjs: A CommonJS specifier. It is not parsed as a URL.
+   *   - url: A URL that works as in a browser. Bare specifiers are treated as relative URLs.
+   *   - custom: A custom specifier. Must be handled by a custom resolver plugin.
+   */
+  +specifierType: SpecifierType;
+  /**
+   * When the dependency should be loaded.
+   *   - sync: The dependency should be resolvable synchronously. The resolved asset will be placed
+   *       in the same bundle as the parent, or another bundle that's already on the page.
+   *   - parallel: The dependency should be placed in a separate bundle that's loaded in parallel
+   *       with the current bundle.
+   *   - lazy: The dependency should be placed in a separate bundle that's loaded later.
+   * @default 'sync'
+   */
+  +priority: DependencyPriority;
+  /**
+   * When the dependency is a bundle entry (priority is "parallel" or "lazy"), this controls the naming
+   * of that bundle. `needsStableName` indicates that the name should be stable over time, even when the
+   * content of the bundle changes. This is useful for entries that a user would manually enter the URL
+   * for, as well as for things like service workers or RSS feeds, where the URL must remain consistent
+   * over time.
+   */
+  +needsStableName: boolean;
+  /** Whether the dependency is optional. If the dependency cannot be resolved, this will not fail the build. */
   +isOptional: boolean;
-  /** Whether an URL is expected (rather than the language-specific behaviour). */
-  +isURL: boolean;
-  +isIsolated: boolean;
-  /** Used for error messages, the code location that caused this dependency. */
+  /** Whether the dependency is an entry. */
+  +isEntry: boolean;
```

Comment:
Ended up needing to keep `isEntry`, but now it actually means that this dependency was an entry rather than anything else (e.g. `needsStableName` is now a separate field). It also cannot be specified by plugins in DependencyOptions, it can only be set by core.